File: programming-language-hw2/problem2/stateful_pipeline.rb
DATA = [3, -1, 4, 0, -2, 5, 8]

def make_greater_than(threshold)
  ->(value) { value > threshold }
end

def make_transformer
  count = 0

  transformer = lambda do |value|
    count += 1
    value * 2
  end

  get_count = -> { count }
  { apply: transformer, get_count: get_count }
end

def process(data, predicate, transformer)
  data.sum do |value|
    predicate.call(value) ? transformer[:apply].call(value) : 0
  end
end

def get_count(transformer)
  transformer[:get_count].call
end

predicate = make_greater_than(2)
transformer = make_transformer

result = process(DATA, predicate, transformer)
count = get_count(transformer)

puts "result = #{result}"
puts "count = #{count}"

File: programming-language-hw2/problem2/stateful_pipeline.py
data = [3, -1, 4, 0, -2, 5, 8]


def make_greater_than(threshold):
    def predicate(value):
        return value > threshold

    return predicate


def make_transformer():
    count = 0

    def transformer(value):
        nonlocal count
        count += 1
        return value * 2

    transformer.get_count = lambda: count
    return transformer


def process(data, predicate, transformer):
    total = 0
    for value in data:
        if predicate(value):
            total += transformer(value)
    return total


def get_count(transformer):
    return transformer.get_count()


predicate = make_greater_than(2)
transformer = make_transformer()

result = process(data, predicate, transformer)
count = get_count(transformer)

print(f"result = {result}")
print(f"count = {count}")
